### PR TITLE
fix loading config from pubspec.yaml

### LINKED ISSUE #399: Unable to run Flutter Launcher Icons without `flutter_launcher_icons.yaml` file
> Hi @RatakondalaArun ! 
> I think I found an issue that can block this release -- I can't run script without having `flutter_launcher_icons.yaml`.
>
> It seems that the reason for this is in the following code:
>https://github.com/fluttercommunity/flutter_launcher_icons/blob/master/lib/main.dart#L184
>```
>// Try pubspec.yaml for compatibility
>    if (configFile == null) {
>      try {
>        return loadConfigFile('pubspec.yaml', fileOptionResult);
>      } catch (_) {}
>    }
>```
>`configFile` can't be `null`, because it has `defaultConfigFile` as a default value.
>
>As as result, I always get `NoConfigFoundException` without `flutter_launcher_icons.yaml` file.
>
>PR: https://github.com/fluttercommunity/flutter_launcher_icons/pull/398

_Originally posted by @p-mazhnik in https://github.com/fluttercommunity/flutter_launcher_icons/issues/397#issuecomment-1203046546_

Note: This is only affecting `master` branch and not a version of Flutter Launcher Icons uploaded to Pub

### PR DESCRIPTION
**Pull request by p-mazhnik**
I found an issue -- can't run script without having `flutter_launcher_icons.yaml`.

It seems that the reason for this is in the following code:
https://github.com/fluttercommunity/flutter_launcher_icons/blob/master/lib/main.dart#L184
```
// Try pubspec.yaml for compatibility
    if (configFile == null) {
      try {
        return loadConfigFile('pubspec.yaml', fileOptionResult);
      } catch (_) {}
    }
```
`configFile` can't be `null`, because it has `defaultConfigFile` as a default value.

As as result, I always get `NoConfigFoundException` without `flutter_launcher_icons.yaml` file.

Resolves #399

### CODE REVIEW

**Comment by p-mazhnik** [comment]
Actually I don't like current solution (and it even breaks tests).
Probably we shouldn't have default value for the `-f` flag, or we should fix the tests and let `pubspec.yml` to be a fallback, if config provided using `-f` flag is not exists

**Comment by RatakondalaArun** [comment]
@p-mazhnik @MarkOSullivan94 

Instead of using `loadConfigFileFromArgResults()` we can use [`FlutterLauncherIconsConfig#toJson()`](https://github.com/fluttercommunity/flutter_launcher_icons/blob/master/lib/flutter_launcher_icons_config.dart#L134-L136) instance method to get already loaded configs

- This prevents us from loading configs twice
- Fallback to `pubspec.yaml` is already included

https://github.com/fluttercommunity/flutter_launcher_icons/blob/4ffd0f544af0b1accaafc686fe392b365b9b43b4/lib/main.dart#L77-L79

**Comment by p-mazhnik** [comment]
@RatakondalaArun I can try to refactor this in the current PR

**Comment by MarkOSullivan94** [comment]
Converting this to draft until changes are made

**Comment by p-mazhnik** [comment]
@RatakondalaArun I couldn't just use `FlutterLauncherIconsConfig#toJson()` because result will always contain all keys, but we had functions like `hasAndroidConfig` that are checked for the keys existence. So I refactored the code to use `FlutterLauncherIconsConfig` instead of `Map<String, dynamic>`

**Comment by RatakondalaArun** [comment]
@p-mazhnik 
I think there is no need for so many changes

1.`createIconsFromArguments()`: Is the entry point. This is where parsing and loading configs happen
2.`createIconsFromConfig()`: Is responsible for checking config to see if configs are present or not (i.e: `hasAndroidConfig` will go inside this function)

Below are the changes that are required prefixed with (👇)
```dart
final ArgResults argResults = parser.parse(arguments);
  // creating logger based on -v flag
  final logger = FLILogger(argResults[verboseFlag]);
  // 👇 Move this to here
  final String prefixPath = argResults[prefixOption]; 

  logger.verbose('Received args ${argResults.arguments}');

  if (argResults[helpFlag]) {
    stdout.writeln('Generates icons for iOS and Android');
    stdout.writeln(parser.usage);
    exit(0);
  }

  // Flavors management
  final flavors = getFlavors();
  final hasFlavors = flavors.isNotEmpty;

  // Load configs from given file(defaults to ./flutter_launcher_icons.yaml) or from ./pubspec.yaml
  final flutterLauncherIconsConfigs =
      FlutterLauncherIconsConfig.loadConfigFromPath(argResults[fileOption], prefixPath) ??
          FlutterLauncherIconsConfig.loadConfigFromPubSpec(prefixPath);

 // 👇 Update this to only check for flutterLauncherIconsConfigs
  if (flutterLauncherIconsConfigs == null) {
    throw NoConfigFoundException(
      'No configuration found in $defaultConfigFile or in ${constants.pubspecFilePath}. '
      'In case file exists in different directory use --file option',
    );
  }

  // 👇 Instead of using loadConfigFileFromArgResults use flutterLauncherIconsConfigs#toJson
  // Load the config file
  final Map<String, dynamic> yamlConfig = flutterLauncherIconsConfigs.toJson(); //loadConfigFileFromArgResults(argResults, verbose: true);
```
@MarkOSullivan94 correct me if I am wrong

**Comment by p-mazhnik** [comment]
@RatakondalaArun 
problem here is that after you will pass new `yamlConfig` into `createIconsFromConfig` function, all the checks won't work as expected, such as `isImagePathInConfig`, `hasPlatformConfig`, `isNeedingNewAndroidIcon`, etc. 
Because `flutterLauncherIconsConfigs.toJson()` will return Map that will contain all the keys, so code like `flutterLauncherIconsConfig.containsKey('ios')` always will return `true`. That's why I made this refactoring.

File: test/android_test.dart [hunk 1]
```diff
@@ -31,14 +32,22 @@ void main() {
       'android': true,
       'ios': true
     };
-    expect(android.isCustomAndroidFile(flutterIconsConfig), false);
+    expect(
+      FlutterLauncherIconsConfig.fromJson(flutterIconsConfig)
+          .isCustomAndroidFile,
+      false,
```
Comment (RatakondalaArun) [inline test/android_test.dart:38]:
```suggestion
      isTrue,
```

File: test/android_test.dart [hunk 2]
```diff
@@ -31,14 +32,22 @@ void main() {
...
     final Map<String, dynamic> flutterIconsNewIconConfig = <String, dynamic>{
       'image_path': 'assets/images/icon-710x599.png',
       'android': 'New Icon',
       'ios': true
     };
-    expect(android.isCustomAndroidFile(flutterIconsNewIconConfig), true);
+    expect(
+      FlutterLauncherIconsConfig.fromJson(flutterIconsNewIconConfig)
+          .isCustomAndroidFile,
+      true,
```
Comment (RatakondalaArun) [inline test/android_test.dart:49]:
```suggestion
      isTrue,
```

File: test/android_test.dart [hunk 3]
```diff
@@ -48,8 +57,11 @@ void main() {
       'android': 'New Icon',
       'ios': true
     };
-    expect(android.getAndroidIconPath(flutterIconsNewIconConfig),
-        'assets/images/icon-android.png');
+    expect(
+      FlutterLauncherIconsConfig.fromJson(flutterIconsNewIconConfig)
+          .getImagePathAndroid(),
+      'assets/images/icon-android.png',
```
Comment (RatakondalaArun) [inline test/android_test.dart:63]:
```suggestion
      equals('assets/images/icon-android.png'),
```

File: test/flutter_launcher_icons_config_test.dart [hunk 4]
```diff
@@ -75,7 +75,31 @@ void main() {
...
       });
     });
-
+    group('#loadConfigFromTestPubSpec', () {
+      test('should return valid configs', () {
+        const String path = 'test/config/test_pubspec.yaml';
+        final configs = FlutterLauncherIconsConfig.loadConfigFromPath(path, '.');
+        expect(configs, isNotNull);
+        const String imagePath = 'assets/images/icon-710x599.png';
+        expect(configs!.imagePath, imagePath);
```
Comment (RatakondalaArun) [inline test/flutter_launcher_icons_config_test.dart:84]:
```suggestion
        expect(configs!.imagePath, equals(imagePath));
```

File: test/main_test.dart [hunk 5]
```diff
@@ -58,10 +58,10 @@ flutter_icons:
   ios: false
 ''');
       final ArgResults argResults = parser.parse(<String>[]);
-      final Map<String, dynamic>? config =
+      final FlutterLauncherIconsConfig? config =
           main_dart.loadConfigFileFromArgResults(argResults);
       expect(config, isNotNull);
-      expect(config!['android'], true);
+      expect(config!.android, true);
```
Comment (RatakondalaArun) [inline test/main_test.dart:64]:
```suggestion
      expect(config!.android, isTrue);
```

File: test/main_test.dart [hunk 6]
```diff
@@ -71,14 +71,14 @@ flutter_icons:
   ios: false
 ''');
       ArgResults argResults = parser.parse(<String>[]);
-      final Map<String, dynamic>? config =
+      final FlutterLauncherIconsConfig? config =
           main_dart.loadConfigFileFromArgResults(argResults);
       expect(config, isNotNull);
-      expect(config!['ios'], false);
+      expect(config!.ios, false);
```
Comment (RatakondalaArun) [inline test/main_test.dart:77]:
```suggestion
      expect(config!.ios, isFalse);
```

File: test/main_test.dart [hunk 7]
```diff
@@ -90,10 +90,10 @@ flutter_icons:
 ''');
       // if no argument set, should fail
       ArgResults argResults = parser.parse(<String>['-f', 'custom.yaml']);
-      final Map<String, dynamic>? config =
+      final FlutterLauncherIconsConfig? config =
           main_dart.loadConfigFileFromArgResults(argResults);
       expect(config, isNotNull);
-      expect(config!['ios'], true);
+      expect(config!.ios, true);
```
Comment (RatakondalaArun) [inline test/main_test.dart:96]:
```suggestion
      expect(config!.ios, isTrue);
```

File: test/main_test.dart [hunk 8]
```diff
@@ -105,32 +105,32 @@ flutter_icons:
...
-
   test('image_path is in config', () {
     final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
       'image_path': 'assets/images/icon-710x599.png',
       'android': true,
       'ios': true
     };
-    expect(main_dart.isImagePathInConfig(flutterIconsConfig), true);
+    final config = FlutterLauncherIconsConfig.fromJson(flutterIconsConfig);
+    expect(config.getImagePathAndroid(), 'assets/images/icon-710x599.png');
```
Comment (RatakondalaArun) [inline test/main_test.dart:115]:
```suggestion
    expect(config.getImagePathAndroid(), equals('assets/images/icon-710x599.png'));
```

File: test/main_test.dart [hunk 9]
```diff
@@ -105,32 +105,32 @@ flutter_icons:
...
+    expect(config.getImagePathIOS(), 'assets/images/icon-710x599.png');
     final Map<String, dynamic> flutterIconsConfigAndroid = <String, dynamic>{
       'image_path_android': 'assets/images/icon-710x599.png',
       'android': true,
       'ios': true
     };
-    expect(main_dart.isImagePathInConfig(flutterIconsConfigAndroid), false);
+    final configAndroid = FlutterLauncherIconsConfig.fromJson(flutterIconsConfigAndroid);
+    expect(configAndroid.getImagePathAndroid(), 'assets/images/icon-710x599.png');
```
Comment (RatakondalaArun) [inline test/main_test.dart:123]:
```suggestion
    expect(configAndroid.getImagePathAndroid(), equals('assets/images/icon-710x599.png'));
```

File: test/main_test.dart [hunk 10]
```diff
@@ -105,32 +105,32 @@ flutter_icons:
...
-      'image_path_android': 'assets/images/icon-710x599.png',
-      'image_path_ios': 'assets/images/icon-710x599.png',
+      'image_path_android': 'assets/images/icon-android.png',
+      'image_path_ios': 'assets/images/icon-ios.png',
       'android': true,
       'ios': true
     };
-    expect(main_dart.isImagePathInConfig(flutterIconsConfigBoth), true);
+    final configBoth = FlutterLauncherIconsConfig.fromJson(flutterIconsConfigBoth);
+    expect(configBoth.getImagePathAndroid(), 'assets/images/icon-android.png');
```
Comment (RatakondalaArun) [inline test/main_test.dart:132]:
```suggestion
    expect(configBoth.getImagePathAndroid(), equals('assets/images/icon-android.png'));
```

File: test/main_test.dart [hunk 11]
```diff
@@ -139,14 +139,16 @@ flutter_icons:
       'android': true,
       'ios': true
     };
-    expect(main_dart.hasPlatformConfig(flutterIconsConfig), true);
+    final config = FlutterLauncherIconsConfig.fromJson(flutterIconsConfig);
+    expect(config.hasPlatformConfig, true);
```
Comment (RatakondalaArun) [inline test/main_test.dart:143]:
```suggestion
    expect(config.hasPlatformConfig, isTrue);
```

File: test/main_test.dart [hunk 12]
```diff
@@ -139,14 +139,16 @@ flutter_icons:
...
   });
 
   test('No platform specified in config', () {
     final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
       'image_path': 'assets/images/icon-710x599.png'
     };
-    expect(main_dart.hasPlatformConfig(flutterIconsConfig), false);
+    final config = FlutterLauncherIconsConfig.fromJson(flutterIconsConfig);
+    expect(config.hasPlatformConfig, false);
```
Comment (RatakondalaArun) [inline test/main_test.dart:151]:
```suggestion
    expect(config.hasPlatformConfig, isFalse);
```

File: test/main_test.dart [hunk 13]
```diff
@@ -155,15 +157,17 @@ flutter_icons:
       'android': false,
       'ios': true
     };
-    expect(main_dart.isNeedingNewAndroidIcon(flutterIconsConfig), false);
+    final config = FlutterLauncherIconsConfig.fromJson(flutterIconsConfig);
+    expect(config.isNeedingNewAndroidIcon, false);
```
Comment (RatakondalaArun) [inline test/main_test.dart:161]:
```suggestion
    expect(config.isNeedingNewAndroidIcon, isFalse);
```

File: test/main_test.dart [hunk 14]
```diff
@@ -155,15 +157,17 @@ flutter_icons:
...
   });
 
   test('No new Android icon needed - no Android config', () {
     final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
       'image_path': 'assets/images/icon-710x599.png',
       'ios': true
     };
-    expect(main_dart.isNeedingNewAndroidIcon(flutterIconsConfig), false);
+    final config = FlutterLauncherIconsConfig.fromJson(flutterIconsConfig);
+    expect(config.isNeedingNewAndroidIcon, false);
```
Comment (RatakondalaArun) [inline test/main_test.dart:170]:
```suggestion
    expect(config.isNeedingNewAndroidIcon, isFalse);
```

File: test/main_test.dart [hunk 15]
```diff
@@ -172,14 +176,16 @@ flutter_icons:
       'android': true,
       'ios': false
     };
-    expect(main_dart.isNeedingNewIOSIcon(flutterIconsConfig), false);
+    final config = FlutterLauncherIconsConfig.fromJson(flutterIconsConfig);
+    expect(config.isNeedingNewIOSIcon, false);
```
Comment (RatakondalaArun) [inline test/main_test.dart:180]:
```suggestion
    expect(config.isNeedingNewIOSIcon, isFalse);
```

**Review by RatakondalaArun — CHANGES_REQUESTED** [review]
Few changes in tests

**Review by RatakondalaArun — APPROVED** [review]
LGTM

**Pull request merged**